applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp

```diff
@@ -32,129 +32,197 @@ namespace Kratos
 
     // making a "time step forwards" and updating the
     const double current_time = mrModelPart.GetProcessInfo()[TIME];
     const double current_dt = mrModelPart.GetProcessInfo()[DELTA_TIME];
-    mQNet2 = mQNet1;
-    mQNet1 = mQNet0;
-    mQNet0 = net_inflow_inlet + net_inflow_outlet;
 
-    // adding time-integrated net inflow ( = new water volume ) to the theoretical value
-    // The value of the integral is evaluated following the idea of the Adams-Moulton-Procedure (s=2)
-    mTheoreticalNegativeVolume += current_dt * ( 5.0*mQNet0 + 8.0*mQNet1 - 1.0*mQNet2 ) / 12.0;
+    mQNet0 = net_inflow_inlet + net_inflow_outlet;
+    mTheoreticalNegativeVolume += current_dt * mQNet0;
 
-    // calculation of the "missing" water volume
-    const double water_volume_error = mTheoreticalNegativeVolume - neg_vol;
 
-    double shift_for_correction = 0.0;
-    // check if it is time for a correction (if wished for)
-    if ( mPerformCorrections && mrModelPart.GetProcessInfo()[STEP] % mCorrectionFreq == 0 && inter_area > 10e-7){
-        // if water is missing, a shift into negative direction increases the water volume
-        shift_for_correction = - water_volume_error / inter_area;
-        ShiftDistanceField( shift_for_correction );
-    }
+    mWaterVolumeError = mTheoreticalNegativeVolume - neg_vol;
 
     // assembly of the log message
     std::string output_line_timestep =  std::to_string(current_time) + "\t\t";
     output_line_timestep +=             std::to_string(neg_vol) + "\t\t";
     output_line_timestep +=             std::to_string(pos_vol) + "\t\t";
-    output_line_timestep +=             std::to_string(water_volume_error) + "\t\t";
+    output_line_timestep +=             std::to_string(mWaterVolumeError) + "\t\t";
     output_line_timestep +=             std::to_string(net_inflow_inlet) + "\t\t";
     output_line_timestep +=             std::to_string(net_inflow_outlet) + "\t\t";
-    output_line_timestep +=             std::to_string(inter_area) + "\t\t";
-    output_line_timestep +=             std::to_string(shift_for_correction) +"\n";
+    output_line_timestep +=             std::to_string(inter_area) + "\n";
+
     return output_line_timestep;
 }
 
 
 
-void MassConservationCheckProcess::ComputeVolumesAndInterface( double& positiveVolume, double& negativeVolume, double& interfaceArea ){
+double MassConservationCheckProcess::ComputeDtForConvection(){
+
+    // a small step is set to avoid numerical problems
+    double time_step_for_convection = 1.0e-7;
+    const auto& r_comm = mrModelPart.GetCommunicator();
+
+    if ( mWaterVolumeError > 0.0 ){
+        // case: water volume was lost by mistake
+        mAddWater = true;
+        double water_outflow_over_boundary = OrthogonalFlowIntoAir( 1.0 );
+        if ( r_comm.GetDataCommunicator().SumAll( water_outflow_over_boundary ) ){
+            // checking if flow is sufficient (avoid division by 0)
+            if ( water_outflow_over_boundary > 1.0e-7 ){
```

Comment:
1e-7 looks big to me just to avoid division by 0.0. In small parts 1e-7 might be not so negligible. 

On the other hand 1e-7 in bigger parts might make no senes. Probably safe to move to a %of the overall volume with some numeric restriction.

I mean if your overall relative error wrt volume is 1e-12 safe and faster to skip.

Reply:
good point